lib/go/thrift/binary_protocol.go

```diff
@@ -482,35 +480,37 @@ func (p *TBinaryProtocol) readAll(ctx context.Context, buf []byte) (err error) {
 const readLimit = 32768
 
 func (p *TBinaryProtocol) readStringBody(size int32) (value string, err error) {
+	buf, err := safeReadBytes(size, p.trans, p.buffer[:], readLimit)
+	return buf.String(), NewTProtocolException(err)
+}
+
+// This function is shared between TBinaryProtocol and TCompactProtocol.
+func safeReadBytes(size int32, trans io.Reader, buffer []byte, singleReadLimit int) (*bytes.Buffer, error) {
```

Comment:
I would rather return just the slice `[]byte`, which would be heap-allocated anyway, and keep the `bytes.Buffer` hidden and allocated on the stack, as it was previously.

Also the semantics of this function is a bit confusing - what is the role of `buffer` argument?

Reply:
Good point. Actually we don't need to specify the limit at all. Combining bytes.Buffer with io.CopyN is enough.